Initialize ktor backend project

zws33 opened the pull request:
Setup crud endpoints for recipes and ingredients
Setup endpoint to get shopping list from list of recipe ids

zws33 committed:
Initialize ktor backend project
Setup crud endpoints for recipes and ingredients
Setup endpoint to get shopping list from list of recipe ids

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/Models.kt:
```diff
@@ -0,0 +1,22 @@
+package me.zwsmith.plugins
```
What's the significance of the `plugins` folder?

Do these models belong in it or a more generic folder?

zws33 replied:
i'm not sure. it's part of the ktor architectural philosophy but i'm not sure what the right way to organize stuff is. https://ktor.io/docs/server-plugins.html

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/Models.kt:
```diff
@@ -0,0 +1,22 @@
+data class Recipe(
+    val id: Int,
+    val name: String,
+    val description: String,
```
This feels like it should be nullable/optional, otherwise you'll have to store blank strings as descriptions and that would lead to annoying validation

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/Models.kt:
```diff
@@ -0,0 +1,22 @@
+data class QuantifiedIngredient(
+    val ingredient: Ingredient,
+    val amount: Float,
+    val unit: String,
```
Can we make the unit an enum? We can do that later too but just a thought

zws33 replied:
Yea I was debating that too. Wasn't sure if that would complicate things more than I needed at this stage. But I'm down to make the adjustment 👍

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/database/RecipeService.kt:
```diff
@@ -0,0 +1,135 @@
+        recipe.toRecipe()
+    }
+    
+    suspend fun updateRecipeIngredients(recipeId: Int, ingredients: List<QuantifiedIngredient>) = dbQuery {
```
This name should be `replaceRecipeIngredients`

I would expect `updateRecipeIngredients` to not delete existing ingredients entirely, but rather update quantities of existing ingredients, and insert ingredients not already present.

zws33 replied:
Ah interesting. I agree. This implementation actually came from not knowing how to use the DAO api to just modify the rows in the `QuantifiedIngredients` table. But this was what I ended up coming with a figured I would just enforce that the incoming ingredient list would be the new source of truth and whatever relations existed were in valid?

yousuf-haque replied:
yeah that's fine, I just think you should update the name to better reflect what's actually happening.

If you need an update ingredients later, you can add it later

zws33 replied:
I was actually able to figure out how to update existing relations and add new ones in the one transaction. No need to delete

zws33 replied:
Not sure if the error handling is optimal though. I wanted it to throw so that it would rollback the transaction if any of the new ingredient relationships passed don't exist

yousuf-haque replied:
Invoke [`rollback`](https://jetbrains.github.io/Exposed/api/exposed-core/org.jetbrains.exposed.sql/-transaction/rollback.html) to abort the transaction instead of throwing

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/database/RecipeService.kt:
```diff
@@ -0,0 +1,135 @@
+    suspend fun findRecipeById(id: Int): Recipe = dbQuery {
+        RecipeDao.findById(id)
+            ?.toRecipe()
+            ?: throw IllegalArgumentException("Recipe with id $id does not exist")
```
I don't think throwing an exception here is the right move - that seems aggressive. Returning null/optional seems more expressive and ergonomic

zws33 replied:
Yea error handling in general has been something I've been unsure about. Like if this was and Android app with my usual unidirectional data flow approach I would just wrap everything in results but was thinking maybe that was over engineering/premature. You think I should update the service methods to expose `Results`?

yousuf-haque replied:
I do think exposing results would be useful, and not premature. It would greatly simply the code you have and make it much easier to write code later

Kotlin has a built in [`Result`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-result/) type now

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/database/RecipeService.kt:
```diff
@@ -0,0 +1,135 @@
+            // Associate QuantifiedIngredients with the new Recipe
+            ingredientInputs.forEach { input ->
+                val ingredient = IngredientDao
+                    .find { Ingredients.name eq input.ingredient.name }
```
is `ingredient.name` uniqueness enforced?

zws33 replied:
Oh woops not in this setup. I need to update the table definition.

yousuf-haque replied:
I ask because if uniqueness is not enforced, you may end up with the wrong ingredient that has the same name (contrived edge case but its best to make invalid states unrepresentable)

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/recipes/RecipeController.kt:
```diff
@@ -0,0 +1,158 @@
+        driver = "org.h2.Driver",
+        password = "",
+    )
+    val recipeService = RecipeService(database)
```
`recipeService` should probablu be passed into this function as a dependency rather than being created here.

Both the database connection, and the service should be instantiated in application/main when you're wiring up the dependencies for the application

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/recipes/RecipeController.kt:
```diff
@@ -0,0 +1,158 @@
+            val id = call.parameters["id"]?.toInt() ?: return@get call.respond(HttpStatusCode.BadRequest)
+            val recipe = try {
+                recipeService.findRecipeById(id)
+            } catch (e: Exception) {
```
Instead of catching an exception for failing to find a recipe, you can check if the found recipe is `null`/`None` and then map that to a `NotFound` status code

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/recipes/RecipeController.kt:
```diff
@@ -0,0 +1,158 @@
+        get("/ingredients/{id}") {
+            val id = call.parameters["id"]?.toInt() ?: return@get call.respond(HttpStatusCode.BadRequest)
+            val ingredient = recipeService.findIngredientById(id)
+            if (ingredient == null) {
```
I think recipe finding should be done exactly like this

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/recipes/RequestData.kt:
```diff
@@ -0,0 +1,34 @@
+import me.zwsmith.plugins.QuantifiedIngredient
+
+@Serializable
+data class RecipeInput(
```
These requests should be colocated with the controller/service defining them, as they are part of their public APIs

yousuf-haque reviewed (approved):
Looks great overall!

zws33 committed:
Update DI, passing service to router
Add logger
Update recipe update method to modify existing ingredients and add new ingredients
Update QuantifiedIngredients table to be composite id table.
Colocate recipe service input types

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/database/Database.kt:
```diff
@@ -0,0 +1,24 @@
+            url = "jdbc:h2:mem:test;DB_CLOSE_DELAY=-1",
+            user = "root",
+            driver = "org.h2.Driver",
+            password = "",
```
You should eventually pull these into env vars

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/database/RecipeService.kt:
```diff
@@ -1,16 +1,21 @@
+import org.slf4j.LoggerFactory
 
 class RecipeService(database: Database) {
+    val logger = LoggerFactory.getLogger(RecipeService::class.java)
```
probably should pass in a logger in the constructor too

yousuf-haque commented on backend-ktor/src/main/kotlin/me/zwsmith/plugins/database/RecipeService.kt:
```diff
@@ -85,39 +90,50 @@ class RecipeService(database: Database) {
-        }
-        recipe.toRecipe()
+        recipeUpdate: RecipeUpdate
+    ): Recipe? = dbQuery {
```
I think its reasonable for this function to return an error result if the recipe id is not found

zws33 committed:
Update error handling to use result class

zws33 committed:
Refactor routing definition
Extract interface from recipe service

zws33 committed:
Add docs with AI assistant
